File: 04-computation/continuing8_20260906_lrc_minimum_tree_audit.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;
using I=long long;
using Word=array<int,7>;
const I INF=10000000000LL;
uint64_t gates=0, unpruned=0, wordchecks=0, evaluations=0, compatible=0;
void need(bool x,const string& s){++gates;if(!x){cerr<<"FAILED "<<s<<"\n";exit(1);}}
I mod(I a,I b){I r=a%b;return r<0?r+b:r;}
I floorq(I a,I b){return (a-mod(a,b))/b;}
I ceilq(I a,I b){return -floorq(-a,b);}
int gd[91][91];
struct Geo{int p,q;I L;vector<pair<I,I>> intervals;};
bool danger(I x,I den,int p){I z=mod(p*x,den);return 14*min(z,den-z)<den;}
Geo geometry(int p,int q){
    Geo G{p,q,14LL*p*q,{}};
    vector<I> walls={0,G.L};
    for(int k=0;k<=p;k++)for(int sign:{-1,1}){I x=(14LL*k+sign)*q;if(0<x&&x<G.L)walls.push_back(x);}
    for(int k=0;k<=q;k++)for(int sign:{-1,1}){I x=(14LL*k+sign)*p;if(0<x&&x<G.L)walls.push_back(x);}
    sort(walls.begin(),walls.end());walls.erase(unique(walls.begin(),walls.end()),walls.end());
    vector<pair<I,I>> cells;
    for(size_t i=1;i<walls.size();i++){
        I a=walls[i-1],b=walls[i];
        if(danger(a+b,2*G.L,p)&&danger(a+b,2*G.L,q)){
            if(!cells.empty()&&cells.back().second==a&&danger(a,G.L,p)&&danger(a,G.L,q))cells.back().second=b;
            else cells.push_back({a,b});
        }
    }
    need(cells.size()>=2&&cells.front().first==0&&cells.back().second==G.L,"raw wall origin components");
    G.intervals.push_back({cells.back().first-G.L,cells.front().second});
    G.intervals.insert(G.intervals.end(),cells.begin()+1,cells.end()-1);
    return G;
}
I count_at(const Geo&G,int n,I phase2){
    I ans=0;for(auto [a,b]:G.intervals)ans+=ceilq(2LL*n*b-phase2,2*G.L)-floorq(2LL*n*a-phase2,2*G.L)-1;return ans;
}
pair<I,I> capacity(const Geo&G,int n){
    map<I,pair<int,int>> events;
    for(auto [a,b]:G.intervals){events[mod(n*a,G.L)].first++;events[mod(n*b,G.L)].second++;}
    I best=count_at(G,n,0),phase=0,cur=count_at(G,n,1);
    if(cur<best){best=cur;phase=1;}
    for(auto [wall,se]:events){
        if(wall==0)continue;
        I at=cur-se.second;
        if(at<best){best=at;phase=2*wall;}
        cur=at+se.first;
        if(cur<best){best=cur;phase=2*wall+1;}
    }
    return {best,phase};
}
I literal(const Geo&G,int n,I phase2){
    I ans=0,den=2*G.L*n;
    for(int j=0;j<n;j++){I x=2*G.L*j+phase2;if(danger(x,den,G.p)&&danger(x,den,G.q))ans++;}
    return ans;
}
vector<Geo> atlas(){
    vector<int> primes;
    for(int p=2;p<=356;p++){bool prime=true;for(int d=2;d*d<=p;d++)if(p%d==0){prime=false;break;}if(prime&&p%3==2)primes.push_back(p);}
    set<int> sums={1};
    for(int p:primes){vector<int> old(sums.begin(),sums.end());for(int s:old){if(s*p<=356)sums.insert(s*p);if((I)s*p*p<=356)sums.insert(s*p*p);}}
    vector<Geo> out;
    for(int s:sums)if(s>=3)for(int p=1;2*p<s;p++)if(gcd(p,s-p)==1)out.push_back(geometry(p,s-p));
    need(out.size()==5855,"multiplicatively generated strict atlas");return out;
}
array<unordered_set<uint64_t>,7> profiles;
vector<int> masks;
uint64_t key(int c,const vector<int>& w){uint64_t k=c;for(int x:w)k=(k<<7)|x;return k;}
bool valid(const Word&w){
    int all=0;for(int x:w)all=gd[all][x];if(all!=1)return false;
    int cache[128]={0};
    auto get=[&](auto&&self,int m)->int{
        if(cache[m])return cache[m];int bit=__builtin_ctz((unsigned)m),rest=m&(m-1);
        return cache[m]=rest?gd[w[bit]][self(self,rest)]:w[bit];
    };
    for(int m:masks){
        int c=get(get,m),k=7-__builtin_popcount((unsigned)m);array<int,6> a{};int z=0;
        for(int j=0;j<7;j++)if(!(m&(1<<j)))a[z++]=gd[c][w[j]];
        sort(a.begin(),a.begin()+k);uint64_t v=c;for(int j=0;j<k;j++)v=(v<<7)|a[j];
        if(!profiles[k].count(v))return false;
    }
    return true;
}
void enumerate(const vector<int>&D,Word&w,int pos,int first,vector<Word>&out){
    if(pos==7){unpruned++;if(valid(w))out.push_back(w);return;}
    for(int j=first;j<(int)D.size();j++){w[pos]=D[j];enumerate(D,w,pos+1,j,out);}
}
void write_words(const vector<Word>&W,const string&path){
    ofstream f(path,ios::binary);f<<'[';bool first=true;
    for(auto&w:W){if(!first)f<<',';first=false;f<<'[';for(int j=0;j<7;j++){if(j)f<<',';f<<w[j];}f<<']';}f<<']';
}
I mst(const Word&w,const array<array<I,91>,91>& weights){
    array<array<I,3>,21> edges;int z=0;
    for(int i=0;i<7;i++)for(int j=i+1;j<7;j++)edges[z++]={weights[w[i]][w[j]],i,j};
    sort(edges.begin(),edges.end());array<int,7> parent; iota(parent.begin(),parent.end(),0);
    auto find=[&](int x){while(parent[x]!=x)x=parent[x];return x;};
    I total=0;int used=0;
    for(auto ed:edges){int a=find(ed[1]),b=find(ed[2]);if(a!=b){parent[a]=b;total+=ed[0];if(++used==6)break;}}
    return total;
}
struct Expected{int t,domain;I words,minimum,ownerE,ownerTree,expected_events,expected_compatible;Word owner;vector<array<I,9>> survivors;map<pair<int,int>,vector<I>> weights;};
int main(int argc,char**argv){
    if(argc!=3){cerr<<"input and output directory required\n";return 2;}
    ifstream in(argv[1]);string dir=argv[2];need(bool(in),"native input open");
    for(int a=0;a<=90;a++)for(int b=0;b<=90;b++)gd[a][b]=gcd(a,b);
    int np;in>>np;for(int j=0;j<np;j++){int k,c;in>>k>>c;vector<int>w(k);for(int&x:w)in>>x;profiles[k].insert(key(c,w));}
    for(int size=6;size>=1;size--)for(int m=1;m<127;m++)if(__builtin_popcount((unsigned)m)==size)masks.push_back(m);
    need(masks.size()==126,"all126 proper nonempty positional subsets");
    int nd;in>>nd;vector<vector<int>> domains(nd);for(auto&D:domains){int k;in>>k;D.resize(k);for(int&x:D)in>>x;}
    int nt;in>>nt;vector<Expected> ex(nt);
    for(auto&E:ex){
        int ns,nw;in>>E.t>>E.domain>>E.words>>E.minimum>>E.ownerE>>E.ownerTree>>E.expected_events>>E.expected_compatible;for(int&x:E.owner)in>>x;
        in>>ns;E.survivors.resize(ns);for(auto&s:E.survivors)for(I&x:s)in>>x;
        in>>nw;for(int j=0;j<nw;j++){int a,b,k;in>>a>>b>>k;vector<I>v(k);for(I&x:v)in>>x;E.weights[{a,b}]=v;}
    }
    need(bool(in),"complete native input");
    auto AT=atlas();
    set<pair<int,int>> atlas_pairs;for(auto&G:AT)atlas_pairs.insert({G.p,G.q});
    for(auto pq:vector<pair<int,int>>{{1,1},{1,13},{1,112},{1,355},{11,263},{23,323},{33,278},{66,289}}){
        auto G=geometry(pq.first,pq.second);
        for(int n:{1,2,7,13,14,28,29,720,968,2485}){
            auto [best,phase]=capacity(G,n);set<I>walls={0};for(auto ab:G.intervals){walls.insert(mod(n*ab.first,G.L));walls.insert(mod(n*ab.second,G.L));}
            I brute=INF;for(I x:walls){brute=min(brute,count_at(G,n,2*x));brute=min(brute,count_at(G,n,2*x+1));}
            need(best==brute,"all-wall direct ceil-floor control");need(best==literal(G,n,phase),"literal strict original-grid at minimizing phase");
        }
    }
    need(capacity(geometry(1,355),2485).first==0,"exact near-resonance endpoint hostile");
    vector<vector<Word>> banks(nd);I distinct=0;
    for(int d=0;d<nd;d++){
        Word w{};enumerate(domains[d],w,0,0,banks[d]);distinct+=banks[d].size();
        write_words(banks[d],dir+"/words_"+to_string(d)+".json");
        cerr<<"DOMAIN "<<d+1<<'/'<<nd<<" valid="<<banks[d].size()<<"\n";
    }
    I minupper=INF,totalwords=0;int excluded=0,hostiles=0;I pairtables=0;
    for(auto&E:ex){
        auto&D=domains[E.domain];auto&bank=banks[E.domain];need((I)bank.size()==E.words,"unpruned full word count at "+to_string(E.t));
        array<array<I,91>,91> W;for(auto&row:W)row.fill(INF);set<int>es;set<int>states(D.begin(),D.end());
        for(int a:D)for(int b:D)es.insert(gcd(a,b));
        I localcompatible=0,localevents=0;
        for(int e:es){int n=E.t/e;for(auto&G:AT){
            int a=e*gcd(n,G.p),b=e*gcd(n,G.q);if(a>b)swap(a,b);if(!states.count(a)||!states.count(b))continue;
            localcompatible++;I sep=0;for(auto [lo,hi]:G.intervals)sep+=floorq((I)n*(hi-lo)-1,G.L);sep*=e;
            if(sep>=W[a][b])continue;
            I v=e*capacity(G,n).first;evaluations++;localevents++;need(sep<=v&&v<=E.t,"separate open-arc floor lower bound");
            if(v<W[a][b])W[a][b]=W[b][a]=v;
        }}
        compatible+=localcompatible;
        if(E.expected_events>=0)need(localevents==E.expected_events,"complete event-evaluation count");
        if(E.expected_compatible>=0)need(localcompatible==E.expected_compatible,"complete compatible-edge count");
        need(E.weights.size()==D.size()*(D.size()+1)/2,"complete forced-pair table");
        for(auto [ab,v]:E.weights){
            int a=ab.first,b=ab.second;pairtables++;need(W[a][b]==v[0],"independent raw-wall minimum pair credit at "+to_string(E.t));
            if(v.size()>1){int e=v[1],p=v[2],q=v[3];auto G=geometry(p,q);int n=E.t/e;
                need(e==gcd(a,b)&&atlas_pairs.count({p,q}),"attainer gcd and exact strict-atlas membership");
                pair<int,int> got={e*gcd(n,p),e*gcd(n,q)};if(got.first>got.second)swap(got.first,got.second);
                need(got==ab,"attainer actual state pair");need(e*capacity(G,n).first==v[0],"attainer count reconstruction");
                I sep=0;for(auto [lo,hi]:G.intervals)sep+=floorq((I)n*(hi-lo)-1,G.L);need(e*sep==v[4],"attainer separate-open-interval column");
                if(E.t==7200||E.t==12672){auto cp=capacity(G,n);need(cp.first==literal(G,n,cp.second),"complete hostile/minimum-clock attainers on literal native grids");}
            }else need(v[0]==INF,"impossible-pair infinity convention");
        }
        I worst=INF;vector<array<I,9>> bad;
        for(auto&w:bank){
            I Ecost=-E.t,res=0;for(int d:w){Ecost+=(I)d*ceilq(E.t,7*d);res+=(I)d*mod(-E.t/d,7);}
            need(7*Ecost==res&&Ecost>=0,"native marginal residue identity");
            I tree=mst(w,W),margin=tree-Ecost;worst=min(worst,margin);wordchecks++;totalwords++;
            if(margin<=0){array<I,9>row{};for(int j=0;j<7;j++)row[j]=w[j];row[7]=Ecost;row[8]=tree;bad.push_back(row);}
        }
        need(worst==E.minimum,"all-word Kruskal minimum at "+to_string(E.t));need(bad==E.survivors,"complete compatible cost survivors at "+to_string(E.t));
        I oe=-E.t;for(int d:E.owner)oe+=(I)d*ceilq(E.t,7*d);
        need(valid(E.owner)&&oe==E.ownerE&&mst(E.owner,W)==E.ownerTree&&E.ownerTree-oe==worst,"producer extremizer independently valid and attaining");
        if(E.t>=12000){need(bad.empty(),"upper clock closes");minupper=min(minupper,worst);excluded++;}else{need(E.t==7200&&bad.size()==15&&worst==-74,"7200 lower-envelope stopping control");hostiles++;}
        cerr<<"CLOCK "<<E.t<<" words="<<bank.size()<<" margin="<<worst<<" survivors="<<bad.size()<<"\n";
    }
    cout<<"ATLAS 5855; PROPER_SUBSETS 126; DOMAIN_COUNT "<<nd<<"\n";
    cout<<"UNPRUNED_WORDS "<<unpruned<<"; DISTINCT_VALID_WORDS "<<distinct<<"; WORD_CLOCK_EVALUATIONS "<<totalwords<<"\n";
    cout<<"PAIR_TABLE_ENTRIES "<<pairtables<<"; COMPATIBLE_ATLAS_EDGES "<<compatible<<"; EXACT_PROFILE_EVALUATIONS "<<evaluations<<"\n";
    cout<<"UPPER_CLOCKS_EXCLUDED "<<excluded<<"; MINIMUM_MARGIN "<<minupper<<"; HOSTILE_7200_COUNT 15; HOSTILE_MARGIN -74\n";
    need(excluded==549&&hostiles==1,"declared complete selected scope");
    cout<<"PASS "<<gates<<" always-active native exact gates\n";
}

File: 04-computation/continuing8_20260906_lrc_minimum_tree_audit.py
#!/usr/bin/env python3
"""Independent complete native referee: raw walls, unpruned words, Kruskal.

No mathematical producer is imported or executed. The companion C++ engine
reconstructs every domain and all pair costs; the JSON certificate is only an
expected answer. Python validates inherited pins and serializes the native input.
"""
from pathlib import Path
from hashlib import sha256
import argparse
import json
import os
import shutil
import subprocess
import sys
import tempfile

sys.stdout.reconfigure(encoding='utf-8', newline='\n')
HERE = Path(__file__).resolve().parent
STEM = 'continuing8_20260906_lrc_minimum_tree_audit'
GATES = 0


def need(ok, why):
    global GATES
    GATES += 1
    if not ok:
        raise ArithmeticError(why)


def canonical(x):
    return json.dumps(x, sort_keys=True, separators=(',', ':')).encode()


def main():
    ap = argparse.ArgumentParser()
    filed = HERE.name == '04-computation'
    ap.add_argument('--root', type=Path, default=HERE.parent if filed else Path('C:/w/s0905'))
    ap.add_argument('--producer', type=Path, default=HERE.parent / '05-knowledge/results' if filed else Path('C:/w/continuing8_20260906_root'))
    ap.add_argument('--work-dir', type=Path, default=Path(tempfile.gettempdir()) / STEM)
    ap.add_argument('--scout', action='store_true', help='historical development comparison only')
    args = ap.parse_args()
    # Pointwise graph controls: different actual trees and every danger subset.
    trees = [[(i, i + 1) for i in range(6)], [(0, i) for i in range(1, 7)],
             [(0, 1), (0, 2), (1, 3), (1, 4), (2, 5), (2, 6)]]
    for tree in trees:
        for mask in range(128):
            edges = sum(bool(mask & (1 << i)) and bool(mask & (1 << j)) for i, j in tree)
            need(edges <= max(mask.bit_count() - 1, 0), 'induced actual forest multiplicity control')
    need(3 > 3 - 1, 'triangle overlap cannot be added as a forest')
    # Invented nonactual edges may have positive weights while the actual path
    # has weight zero. Their possible maximum is not a lower bound for it.
    zero_path = set(trees[0])
    invented_star_weight = sum(0 if (0, j) in zero_path else 10 for j in range(1, 7))
    need(invented_star_weight == 50 and sum(0 for _ in zero_path) == 0,
         'maximum possible tree is not actual-tree credit')
    profile = args.root / '04-computation/overnight12_20260906_lrc_decoder_descent_inherited_profiles.json'
    baseline = args.root / '05-knowledge/results/continuing7_20260906_lrc_residue_domains_certificate.json'
    raw = profile.read_bytes()
    need(sha256(raw).hexdigest() == '935f3f687b6d7c89cc099e536f536238fd753bcc4c1747906d213cef387ca93f', 'raw complete-profile pin')
    P = json.loads(raw)['levels']
    states = sorted({row[0] for row in P['6']['profiles']})
    need(states == P['6']['gcds'] and len(states) == 42, 'complete sheet alphabet')
    old = json.loads(baseline.read_bytes())['new_scales']
    need(sha256(canonical(old)).hexdigest() == 'c67f5e98eff3fe406a4416057c6063095290330a50f039e731ced0fc2ca4657a', 'complete baseline semantic pin')
    upper = sorted((t for t in old if t >= 12000), reverse=True)
    need(len(old) == 8195 and len(upper) == 549 and 7200 in old, 'complete inherited target selection')
    if args.scout:
        records = [json.loads(s) for s in (args.producer / 'range_12000.jsonl').read_text().splitlines()]
        hostile = json.loads((args.producer / 'forest_7200.json').read_bytes())
        worst = hostile['worst']
        records.append(dict(t=7200, domain=hostile['domain'], word_count=hostile['layers'][-1],
                            minimum_margin=worst[0], owner=worst[1], owner_E=worst[2], owner_tree=worst[3],
                            survivors=hostile['survivors'], weights=hostile['weights']))
        bank_pins = {}
        for f in args.producer.glob('domain_*.json'):
            J = json.loads(f.read_bytes())
            bank_pins[tuple(J['domain'])] = sha256(canonical(J['words'])).hexdigest()
        certificate_sha = 'SCOUT_ONLY'
    else:
        cert_path = args.producer / 'continuing8_20260906_lrc_minimum_tree_certificate.json'
        cert_raw = cert_path.read_bytes()
        certificate_sha = sha256(cert_raw).hexdigest()
        need(certificate_sha == '580a7c930103aab3bea867ad463a90b0e0208323a90ee95a685ff811a761582d', 'frozen complete producer-certificate pin')
        J = json.loads(cert_raw)
        records = J['clocks']
        bank_pins = {tuple(d['domain']): d['words_sha256'] for d in J['domains']}
        need(J['profile_sha256'] == sha256(raw).hexdigest(), 'declared profile dependency')
        need(J['baseline_semantic_sha256'] == sha256(canonical(old)).hexdigest(), 'declared baseline dependency')
        new = [t for t in old if t not in set(upper)]
        need(J['removed_scales'] == upper and J['new_scales'] == new, 'literal complete deletion and retained array')
        need(J['new_scales_sha256'] == sha256(canonical(new)).hexdigest(), 'retained array semantic digest')
    need([r['t'] for r in records] == upper + [7200], 'exact clock order, no missing selected clock')
    domains = sorted({tuple(d for d in states if t % d == 0) for t in upper + [7200]})
    if args.scout:
        bank_pins = {d: bank_pins.get(d) for d in domains}
    need(set(domains) == set(bank_pins), 'every domain and no extraneous domain')
    index = {d: i for i, d in enumerate(domains)}
    lines = []
    profiles = [(int(k), c, w) for k, row in P.items() for c, w in row['profiles']]
    lines.append(str(len(profiles)))
    lines.extend(' '.join(map(str, [k, c] + w)) for k, c, w in profiles)
    lines.append(str(len(domains)))
    lines.extend(' '.join(map(str, [len(d)] + list(d))) for d in domains)
    lines.append(str(len(records)))
    for r in records:
        D = tuple(d for d in states if r['t'] % d == 0)
        need(list(D) == r['domain'], 'native divisibility determines full domain')
        if not args.scout:
            table = {tuple(ab): v[0] for ab, v in r['weights']}
            parents = list(range(7))
            def find(i):
                while parents[i] != i:
                    i = parents[i]
                return i
            total = 0
            need(len(r['owner_edges']) == 6, 'six positional owner-tree edges')
            for i, j, weight in r['owner_edges']:
                need(0 <= i < j < 7 and find(i) != find(j), 'owner edges form an actual positional tree')
                parents[find(i)] = find(j)
                need(weight == table[tuple(sorted((r['owner'][i], r['owner'][j])))], 'owner edge references reconstructed pair table')
                total += weight
            need(total == r['owner_tree'], 'owner edge total')
        lines.append(' '.join(map(str, [r['t'], index[D], r['word_count'], r['minimum_margin'], r['owner_E'], r['owner_tree'], r.get('event_evaluations', -1), r.get('compatible_atlas_edges', -1)] + r['owner'])))
        lines.append(str(len(r['survivors'])))
        lines.extend(' '.join(map(str, word + [excess, credit])) for word, excess, credit in r['survivors'])
        lines.append(str(len(r['weights'])))
        lines.extend(' '.join(map(str, ab + [len(v)] + v)) for ab, v in r['weights'])
    work = args.work_dir / ('optimized' if sys.flags.optimize else 'normal')
    work.mkdir(parents=True, exist_ok=True)
    native_input = work / 'input.txt'
    native_input.write_bytes(('\n'.join(lines) + '\n').encode())
    cpp = HERE / (STEM + '.cpp')
    compiler = shutil.which('g++') or 'C:/Users/Eliott/scoop/apps/gcc/current/bin/g++.exe'
    exe = work / ('referee.exe' if os.name == 'nt' else 'referee')
    options = ['-std=c++17', '-O3', '-DNDEBUG'] if sys.flags.optimize else ['-std=c++17', '-O2']
    subprocess.run([compiler, *options, str(cpp), '-o', str(exe)], check=True, capture_output=True)
    env = os.environ.copy()
    env['PATH'] = str(Path(compiler).resolve().parent) + os.pathsep + env.get('PATH', '')
    with (work / 'progress.txt').open('wb') as err:
        proc = subprocess.run([str(exe), str(native_input), str(work)], stdout=subprocess.PIPE, stderr=err, env=env)
    if proc.returncode:
        raise ArithmeticError((work / 'progress.txt').read_text()[-4000:])
    native = proc.stdout.replace(b'\r\n', b'\n')
    for i, D in enumerate(domains):
        words_raw = (work / ('words_' + str(i) + '.json')).read_bytes()
        if bank_pins[D] is not None:
            need(sha256(words_raw).hexdigest() == bank_pins[D], 'complete unpruned ordered word-list digest')
    new = [t for t in old if t not in set(upper)]
    print('INDEPENDENT_ENGINE raw spatial cells; signed ceil-floor phase sweep; unpruned126-profile census; Kruskal')
    print(native.decode().strip())
    print('RETAINED_SCALES', len(new), 'MAXIMUM', max(new))
    print('RETAINED_SEMANTIC_SHA256', sha256(canonical(new)).hexdigest())
    print('PRODUCER_CERTIFICATE_SHA256', certificate_sha)
    print('PASS', GATES, 'always-active Python input/completeness gates; output LF')


if __name__ == '__main__':
    main()
